be/src/util/radix_sort.h

```diff
@@ -0,0 +1,309 @@
+    using CountType = uint32_t;
+
+    /// The type to which the key is transformed to do bit operations. This UInt is the same size as the key.
+    using KeyBits = conditional_t<sizeof(Key) == 8, uint64_t, uint32_t>;
+
+    static constexpr size_t PART_SIZE_BITS = 8;    /// With what pieces of the key, in bits, to do one pass - reshuffle of the array.
+
+    /// Converting a key into KeyBits is such that the order relation over the key corresponds to the order relation over KeyBits.
+    using Transform = RadixSortFloatTransform<KeyBits>;
+
+    /// An object with the functions allocate and deallocate.
+    /// Can be used, for example, to allocate memory for a temporary array on the stack.
+    /// To do this, the allocator itself is created on the stack.
+    using Allocator = RadixSortMallocAllocator;
+
+    /// The function to get the key from an array element.
+    static Key & extractKey(Element & elem) { return elem; }
+
+    /// Used when fallback to comparison based sorting is needed.
+    /// TODO: Correct handling of NaNs, NULLs, etc
+    static bool less(Key x, Key y) {
+        return x < y;
+    }
+};
+
+
+template <typename KeyBits>
+struct RadixSortIdentityTransform {
+    static constexpr bool transform_is_simple = true;
+
+    static KeyBits forward(KeyBits x)     { return x; }
+    static KeyBits backward(KeyBits x)    { return x; }
+};
+
+
+template <typename TElement>
+struct RadixSortUIntTraits {
+    using Element = TElement;
+    using Key = Element;
+    using CountType = uint32_t;
+    using KeyBits = Key;
+
+    static constexpr size_t PART_SIZE_BITS = 8;
+
+    using Transform = RadixSortIdentityTransform<KeyBits>;
+    using Allocator = RadixSortMallocAllocator;
+
+    static Key & extractKey(Element & elem) { return elem; }
+
+    static bool less(Key x, Key y) {
+        return x < y;
+    }
+};
+
+
+template <typename KeyBits>
+struct RadixSortSignedTransform
+{
+    static constexpr bool transform_is_simple = true;
+
+    static KeyBits forward(KeyBits x)     { return x ^ (KeyBits(1) << (sizeof(KeyBits) * 8 - 1)); }
+    static KeyBits backward(KeyBits x)    { return x ^ (KeyBits(1) << (sizeof(KeyBits) * 8 - 1)); }
+};
+
+
+template <typename TElement>
+struct RadixSortIntTraits {
+    using Element = TElement;
+    using Key = Element;
+    using CountType = uint32_t;
+    using KeyBits = make_unsigned_t<Key>;
+
+    static constexpr size_t PART_SIZE_BITS = 8;
+
+    using Transform = RadixSortSignedTransform<KeyBits>;
+    using Allocator = RadixSortMallocAllocator;
+
+    static Key & extractKey(Element & elem) { return elem; }
+
+    static bool less(Key x, Key y) {
+        return x < y;
+    }
+};
+
+
+template <typename T>
+using RadixSortNumTraits =
+    conditional_t<std::is_integral<T>::value,
+        conditional_t<std::is_unsigned<T>::value,
+            RadixSortUIntTraits<T>,
+            RadixSortIntTraits<T>>,
+        RadixSortFloatTraits<T>>;
+
+
+template <typename Traits>
+struct RadixSort {
```

Comment:
should add some comment to describe how to use it.

Reply:
done